System/Posix/DynamicLinker/Prim.hsc

```diff
@@ -84,10 +85,17 @@ data RTLDFlags
   | RTLD_LOCAL
     deriving (Show, Read)
 
+#if defined(HAVE_DLFCN_H)
+foreign import capi unsafe "dlfcn.h dlopen" c_dlopen :: CString -> CInt -> IO (Ptr ())
```

Comment:
The `dlopen` call can be comparatively expensive.  One can reasonably argue that this should be `safe` without a significant performance penalty.  Once the object is loaded, the other calls should be quick.

Of course the same would then apply to the `ccall` version.

Reply:
I agree that `dlopen` should be `safe` and I would be happy to push another commit to do that. But I think this is part of a wider problem with this package (see #34) and I don't think we should tackle the problem in a piecemeal fashion. So I would prefer to not do this as part of this PR.

Reply:
> I agree that `dlopen` should be `safe` and I would be happy to push another commit to do that. But I think this is part of a wider problem with this package (see #34) and I don't think we should tackle the problem in a piecemeal fashion. So I would prefer to not do this as part of this PR.

I can accept that I guess, though I tend to take a more opportunistic approach. Big bang fixes can be difficult, but spot improvements while looking at adjacent code are sometimes a good way to make progress.  I'll leave the judgement call to you and others.

Reply:
I think "this is bad elsewhere too" isn't a good argument against fixing. 

Rather fix it locally here and open a new ticket to fix the rest as well, unless this is controversial.

Reply:
@vdukhovni @hasufell I pushed added an additional commit that switches the imports of `dlopen` & `dlclose` to `safe`. The reason that `dlopen` needs `safe` is obvious; it interacts with the local file system. But `dlclose` also has to be `safe` because it might trigger the execution of destructors defined in the library.